File: solution1.py
# Time Complexity : O(log N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : no

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if nums == None or len(nums) == 0:
            return [-1, -1]
        n = len(nums) - 1
        first = self.binarySearchFirst(nums, 0, n, target)
        if first == -1:
            return [-1, -1]
        last = self.binarySearchLast(nums, 0, n, target)
        return [first, last]

    def binarySearchFirst(self, nums, low, high, target):
        while low <= high:
            mid = low + (high - low) // 2
            if nums[mid] == target:
                if mid == 0 or nums[mid - 1] < nums[mid]:
                    return mid
                else:
                    high = mid - 1
            elif nums[mid] > target :
                high = mid - 1
            else:
                low = mid + 1
        return -1
    
    def binarySearchLast(self, nums, low, high, target):
        while low <= high:
            mid = low + (high - low) // 2
            if nums[mid] == target:
                if mid == high or nums[mid + 1] > nums[mid]:
                    return mid
                else:
                    low = mid + 1
            elif nums[mid] > target :
                high = mid - 1
            else:
                low = mid + 1
        return -1

File: Solution1.java
class Solution1 {
    public int[] searchRange(int[] nums, int target) {
        int[] result = {-1, -1};
        result[0] = find_first_occurence(nums, target);
        result[1] = find_last_occurence(nums, target);
        return result;
    }

    public int find_first_occurence(int[] nums, int target){
        int low = 0;
        int high = nums.length - 1;
        int index = -1;
        while (low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] < target){
                low = mid + 1;
            }
            else if(nums[mid] > target){
                high = mid - 1;
            }
            else{
                index = mid;
                if(mid > 0){
                    high = mid - 1;
                }
                else{
                    return index;
                }
            }
        }
        return index;
    }

    public int find_last_occurence(int[] nums, int target){
        int low = 0;
        int high = nums.length - 1;
        int index = -1;
        while (low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] < target){
                low = mid + 1;
            }
            else if(nums[mid] > target){
                high = mid - 1;
            }
            else{
                index = mid;
                if(mid < nums.length - 1){
                    low = mid  + 1;
                }
                else{
                    return index;
                }
            }
        }
        return index;
    }
}